WIP: Security and Privacy of the Lightning network

seresistvanandras opened the pull request:
This is a work in progress PR. We (@s-tikhomirov, @carlaKC and me) plan to complete this chapter by mid/late November. Any suggestions, comments or feedback are welcome!

seresistvanandras committed:
Initial commit of the S&P chapter

s-tikhomirov committed:
add section on probing

add linebreak before unnumbered list

add (short) section on payment amounts

add section on definitions of privacy

add section on anonymity set

fix a few typos

add section on privacy in LN vs Bitcoin

seresistvanandras committed:
Added "Linking senders and receivers"

Added "Cross-Layer Deanonymization" section

Started the Topology-related section

carlaKC commented:
Related to #400

carlaKC committed:
security: add section on denial of service attacks

seresistvanandras committed:
Typo fixes, Anon-set Figure, Economic incentives

carlaKC committed:
security: add practical considerations

s-tikhomirov committed:
do editing pass; URLs to asciidoc format

s-tikhomirov commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,597 @@
+
+* No risk: do not accept any incoming channels;
+* Low risk: only accept channels from nodes that have been present in the graph for a longer period, and have some long-lived channels;
+* Medium risk: only accept channels from
```
from whom?

seresistvanandras committed:
Nits, fixes and some words about centralization

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+= Definitions of privacy =
+
+Privacy is a complex topic.
```
Great paragraph!

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+We also assume a degree of connectivity in the network, namely, that Lightning channels form a connected graph (therefore, it is possible to find a path from any sender to any receiver) and that network messages get propagated within certain timeouts.
+
+Here are some possible models of adversaries in the Lightning Network.
+A regular "honest-but-curious" forwarding node can observe payment amounts, the immediately preceding and following nodes, the graph of announced channels with their capacities.
```
nit: "payment amounts" is a bit misleading here IMO, because there's a difference between the amount paid and the amount forwarded through a specific hop along the route.

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Here are some possible models of adversaries in the Lightning Network.
+A regular "honest-but-curious" forwarding node can observe payment amounts, the immediately preceding and following nodes, the graph of announced channels with their capacities.
+A very well-connected node can do the same but to a larger extent.
+For example, consider the developers of a popular wallet who maintain a node that their users connect to by default.
+This node would be responsible for routing a large share of payments to and from the users of that wallet.
```
The thought introduced in the previous paragraph isn't really concluded here - routing a large share of payments ... therefore can observe more? I think it's also safe to mention the single private channel that most wallets have by default.

Roasbeef replied:
I think there's an implicit/missing conclusion here? Seems y'all intend to convey that this wallet author learns information about the payments/patterns of each user that they serve?

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+The important properties of the Lightning Network are mostly centered around confidentiality and availability.
+Here are some of the properties we may want to ensure:
+
+* nobody except the sender and the receiver knows the payment amount
```
should these be capitalized?

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+* nobody except the sender and the receiver knows the payment amount
+* nobody can link senders and receivers
+* nobody can block an honest user from sending and receiving payments
```
nit: if we are going to refer to CIA here, then mention HMAC/that people can't meddle with the integrity of payments/messages that nodes send

Roasbeef replied:
Super strong statement, and I think just given the availability requirements of channels, this only applies to _direct_ channels in the system (Alice and Bob are connected with no one in between them).

nopara73 replied:
This third one belongs to availability, while the first two to confidentiality, so I think these should not be in the same list.

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+What is their anonymity set from your point of view?
+If you don't know them personally, and without any additional information, their anonymity set roughly equals the city's population, including travelers.
+If you additionally consider their appearance, you may be able to roughly estimate their age and exclude the city residents who are obviously older or younger than the person in question from the anonymity set.
+Furthermore, if you notice that the person walks into the office of Company X using an electronic badge, the anonymity set shrinks to the number of Company X's employees and visitors.
```
Awesome explanation! I think you can safely make up a company name (maybe StarBlocks is already mentioned in this book)?

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+Bitcoin was initially perceived as an anonymous currency.
+Indeed, Bitcoin addresses are not linked to real identities.
+However, all transactions are broadcast in cleartext and can be analyzed.
```
Would it be too technical to mention that base layer has had its own privacy issues with transaction broadcast patterns?

nopara73 replied:
I think that's just a random detail there as there are many off-chain privacy leaks those are equally as important or even more important, like sending xpubs to servers.

Maybe what should be done here rather is to make clear the difference between the inherent privacy issues of transactions being public and the additional privacy issues that the Bitcoin ecosystem introduced on top of that.

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Multiple companies deanonymize users of Bitcoin and other cryptocurrencies.
+
+At first glance, Lightning provides better privacy than Bitcoin because Lightning payments are not broadcast to the whole network.
+While this improves the privacy baseline, other properties of the Lightning protocol may make anonymous payments more challenging.
+For instance, larger payments have fewer routing options.
```
Would you say this is a property of the protocol or of the current graph structure?

nopara73 replied:
@carlaKC You suggested in https://github.com/lnbook/lnbook/pull/524/files#r525399837:

> Would it be too technical to mention that base layer has had its own privacy issues with transaction broadcast patterns?

For that I may ask: would you say this is a property of the protocol or of the current usage patterns?

IMO my solution: "differentiating between inherent protocol privacy issues and additional privacy issues" should be also applied here: https://github.com/lnbook/lnbook/pull/524/files#r541590406

The thing is, discussing LN and on-chain privacy in isolation is fine for an academic paper as the models are hard to untangle with all the variables included, however a book should attempt to take a holistic approach and consider all the variables.

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+At first glance, Lightning provides better privacy than Bitcoin because Lightning payments are not broadcast to the whole network.
+While this improves the privacy baseline, other properties of the Lightning protocol may make anonymous payments more challenging.
+For instance, larger payments have fewer routing options.
+This may allow an adversary who controls well-capitalized nodes to route most large payments and discover payment amounts and probably other details.
```
"probably other details" is quite vague?

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+An attacker might be interested in learning the sender and/or the receiver of a payment to reveal certain economic relationships.
+This breach of privacy could harm censorship resistance, as an intermediary node could censor payments to or from certain receivers or senders.
+Ideally, linking senders to receivers should not be possible to peers other than the sender and the receiver.
```
Arguably, the ideal would be that the sender and receiver don't even know each other? 
✨  insert rendez-vous routing magic here ✨

Roasbeef replied:
Yep, such techniques can allow us to achieve sender-receiver privacy (sender doesn't know the true node ID of the receiver, the reverse already exists as is).

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+Now, we turn our attention to an *on-path adversary*.
+Such an adversary might seem convoluted.
+However, the single most central node can already observe close to 50% of all LN payments, while the four most central nodes observe an average of 72% payments.
```
Flatten these into one sentence?

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+In that case, Alice is unable to forward the payment and returns the "insufficient balance" error to Mallory.
+Mallory updates her estimation from "between zero and 1 million" to "between zero and 500 thousand."
+
+Note that in any case, Mallory's estimation becomes twice as precise after just one probing!
```
This is a great paragraph 🕺

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Remember that regular nodes don't know balance distributions in remote channels.
+Therefore, payments can (and often do) fail because of insufficient balance at an intermediary hop.
+Error messages allow the sender to exclude the failing channel from consideration when constructing another route.
+A popular Lightning wallet Zap even performs probing internally to check whether a constructed route can really handle a payment.
```
Safe to say that many applications do this, it's not just zap

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+References:
+
+* Mizrahi, A., Zohar, A. https://arxiv.org/abs/2002.06564[Congestion Attacks in Payment Channel Networks]
```
I think there's another paper I need to link here, anybody remember one specifically about liquidity lockup?

carlaKC commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+While the public announcement of IP addresses may be unavoidable for those nodes that wish to have incoming channels in the LN, linkability across nodes of the same user can be mitigated if the clients for each node are hosted with different service providers and thus IP addresses.
+
+*Cross-Layer Linking: Lightning Nodes and Bitcoin Entities*
+Associating LN nodes to Bitcoin entities is a serious breach of privacy that is exacerbated by the fact that most LN nodes publicly expose their IP addresses.
```
Isn't the biggest threat here eclipse attacks on the bitcoin node?

carlaKC reviewed (commented):
Nice to read all of this coming together!

seresistvanandras commented:
Dear authors: @renepickhardt, @Roasbeef and @aantonop! We are ready with our first feature-full draft. We would greatly appreciate if you could have a review of this chapter. Please let us know your thoughts, comments, critique! :)

aantonop commented:
We are reviewing! Thank you @seresistvanandras and others

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Bitcoin offers individuals the possibility of storing and transferring their wealth without restrictions.
+
+Lightning continues this mission.
+It aims at providing a better user experience with instant payments while not sacrificing censorship resistance.
```
I think this sentence is a bit too strong to present without first attempting to define what is meant by "censorship resistance" here. As it's a layer two system, it makes certain tradeoffs at a fundamental level in favor of faster lower cost payments compared to the based layer. As a result, it makes addition assumptions which may be characterized as it having less "censorship resistance" compared to the base layer.

nopara73 replied:
```suggestion
Unlike trivial scaling solutions: custodial Bitcoin banks, LN aims to scale Bitcoin without compromising on self-custody, which leads to greater censorship resistance in the Bitcoin ecosystem.
```

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Then, they describe the relevant properties of the system and check whether it conforms to the requirements.
+A security model is based on a set of underlying _security assumptions_.
+In cryptographic systems, they are often centered around the mathematical properties of the cryptographic primitives such as ciphers, signatures, and hash functions.
+The security assumptions of the Lightning Network are that the ECDSA signatures, SHA-256 hash function, and other cryptographic functions used in the protocol behave within their security definitions.
```
Only an intro paragraph, but I think one thing worth mentioning here is that since LN is a _networked_ protocol, any security model must also factor in how interaction with the network and also topological constraints affect the privacy of the network. As an example, since we use an onion routing protocol for HTLC extension, we by default don't assume "protection" against a global passive adversary, as such an adversary can use packet+timing information to deduce payment flows.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+We also assume a degree of connectivity in the network, namely, that Lightning channels form a connected graph (therefore, it is possible to find a path from any sender to any receiver) and that network messages get propagated within certain timeouts.
+
+Here are some possible models of adversaries in the Lightning Network.
+A regular "honest-but-curious" forwarding node can observe payment amounts, the immediately preceding and following nodes, the graph of announced channels with their capacities.
```
missing an "and" before the last "the" here?

Roasbeef replied:
Worth mentioning here that with MPP payment amounts are slightly more obfuscated given non uniform split sizes, and also that tech like multi-hop locks also removes the payment ID correlation as well.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+For example, consider the developers of a popular wallet who maintain a node that their users connect to by default.
+This node would be responsible for routing a large share of payments to and from the users of that wallet.
+An even stronger security model implies that multiple nodes are under adversarial control.
+If it so happens that two colluding nodes are on the same payment path, they would understand that they are forwarding HTLCs belonging to the same payment, as such HTLCs have the same payment hash.
```
See comment re payment amts and multi-hop locks above.

nopara73 replied:
"if it so" is strange

```suggestion
If two colluding nodes are on the same payment path, they would understand that they are forwarding HTLCs belonging to the same payment, as such HTLCs have the same payment hash.
```

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Availability ensures that users may expect the system to be functioning most of the time.
+
+The important properties of the Lightning Network are mostly centered around confidentiality and availability.
+Here are some of the properties we may want to ensure:
```
Who's "we" here, it reads as if the authors are attempting to retroactively apply their own security models to LN, rather than attempt to analyze the details of the system and the resulting emerging security attributes of it.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+The important properties of the Lightning Network are mostly centered around confidentiality and availability.
+Here are some of the properties we may want to ensure:
+
+* nobody except the sender and the receiver knows the payment amount
```
These statements all feel rather strong, though I think the context here is that these are just _hypothetical_ properties we may want to have in an ideal payment system?

nopara73 replied:
My reading was it's the aim of a system and not the properties of the discussed system.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+For each privacy goal and security model, there is a certain probability that an attacker succeeds.
+This probability depends on various factors, such as the size and the structure of the network.
+Other things equal, it is generally easier to attack a small network rather than a large one.
```
All things equal?

nopara73 replied:
I think not, that changes the meaning. But I agree that it's strange phrasing.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+=== Differences between LN and Bitcoin in terms of privacy
+
+Bitcoin was initially perceived as an anonymous currency.
```
Was it? Very strong statement to make here, Bitcoin has no explicit anonymity enhancing properties.

nopara73 replied:
ACK, even the comments under the first announcement on the cypherpunk mailing list were concerned about the weak privacy of this system.

seresistvanandras replied:
fixed. added reference

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+At first glance, Lightning provides better privacy than Bitcoin because Lightning payments are not broadcast to the whole network.
+While this improves the privacy baseline, other properties of the Lightning protocol may make anonymous payments more challenging.
+For instance, larger payments have fewer routing options.
```
may have* 

As it depends on the topology of the graph itself and also the make up of the channels connecting them.

seresistvanandras replied:
fixed

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Another relevant difference between Lightning and Bitcoin is that Lightning nodes maintain a permanent identity, whereas Bitcoin nodes do not.
+A Bitcoin user can easily switch nodes used to receive blockchain data and broadcast transactions.
+A Lightning user, on the contrary, sends and receives payments through the nodes they have opened channels with.
+Moreover, the Lightning protocol assumes that routing nodes announce their IP address in addition to their node ID.
```
Worth mentioning that Tor is available, so IP addrs are not required.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+A Lightning user, on the contrary, sends and receives payments through the nodes they have opened channels with.
+Moreover, the Lightning protocol assumes that routing nodes announce their IP address in addition to their node ID.
+This creates a permanent link between node IDs and IP addresses, which may be dangerous, considering that an IP address is often an intermediary step in anonymity attacks linked to the user's physical location and, in most cases, real-world identity.
+It is possible to use Lightning over Tor, but many nodes do not use this functionality.
```
I think this statement needs to be cited, from my PoV I see Tor usage rising by the week for new nodes.

nopara73 replied:
According to https://1ml.com/statistics: 

- Number of Nodes: 15,079
- Tor Onion Service Nodes: 2,715

seresistvanandras replied:
Added a reference to https://1ml.com/statistics

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Similar logic applies to payment receivers.
+Many users open only a handful of payment channels, therefore, limiting their anonymity sets.
+Moreover, in Lightning, the anonymity set is static or at least slowly changing.
+In contrast, one can achieve significantly larger anonymity sets in on-chain CoinJoin transactions.
```
Seems like a bit of an apples-and-oranges comparison: the off-chain layer can employ techniques such a probing and circular payments to self to insert noise into its payment pattern which allows it to passively increase maintain its anonymity set compared to techniques employed using the persistent on-chain transaction graph.

nopara73 replied:
Yup, I think my criticism is aligned with that: https://github.com/lnbook/lnbook/pull/524/files#r541603355

Coinjoins are more fit for an entropist analysis than onion routing.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+.The anonymity set of Alice and Bob constitutes their neighbors.
+image:images/anon-set.png[]
+
+Finally, Lightning users can also be denied service, having their channels blocked by the attacker.
```
Unclear at which level this "blocking" would occur: on-chain, refuse of update, HTLC jamming, etc?

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Finally, Lightning users can also be denied service, having their channels blocked by the attacker.
+Forwarding payments requires capital - a scarce resource! - to be temporarily blocked in in-flight HTLCs along the route.
+An attacker may send many payments but fail to finalize them, occupying honest user's capital for long periods.
+This attack vector is not present (or at least not as obvious) in Bitcoin.
```
An analogous situation would be a user being unable to transact due to full blocks and/or prohibitive fees.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+In summary, while some aspects of the Lightning Network's architecture suggest that it is a step forward in terms of privacy compared to Bitcoin, other properties of the protocol may make attacks on privacy easier.
+Thorough research is needed to evaluate what privacy guarantees the LN provides and improve the state of affairs.
+We are happy to report that multiple researchers teams are working on Lightning privacy.
+In part, this chapter summarizes the research papers available at the time of writing in late 2020.
```
I think this section is missing the clear (imo) advantage privacy wise: you as a user _select_ the set of nodes you wish to interact with in order to send payments, and the transcript of your payments (the partial information) is known only to those nodes. When transacting on-chain, the transaction graph is made public for all of time, and you don't necessarily select which nodes hear about your transactions, or which miners mined the transactions. 

Consider the following statement: so far in 2020, we know _exactly_ how many Bitcoin transactions have happend off-chain. However, no one is able to make a strong claim w.r.t the number of LN transactions that have happened off-chain.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+One of the goals for a privacy-preserving payment system is to hide the payment amount from uninvolved parties.
+The Lightning Network is an improvement over layer-one in this regard.
+While Bitcoin transactions are broadcast in cleartext and can be observed by anyone, Lightning payments only travel through a few nodes along the payment path.
+However, intermediary nodes do see the payment amount.
```
Worth mentioning here that due to MPP, the payment amount they see may not necessarily be the actual total payment amount.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+In the first step of this attack scenario, a potent off-path adversary deduces the individual balances in each payment channel via probing (described in a subsequent section) and forms a network snapshot at time _t_.
+It then runs the attack sometime later at time _t'_ and uses the differences between the two snapshots to infer information about payments that took place by looking at paths that have changed.
+In the simplest case, if only a single payment occurred between time _t'_ and _t_, the adversary observes a single path where the balances have changed by the same amounts.
+Thus, the adversary learns everything about this payment: the sender, the recipient, and the amount.
```
Though it's a simple model, I think this should also mention the existence of unadvertised channels as well, and the potential that a "shadow" network exists beyond the graph view of this attacker.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+Now, we turn our attention to an *on-path adversary*.
+Such an adversary might seem convoluted.
+However, the single most central node can already observe close to 50% of all LN payments, while the four most central nodes observe an average of 72% payments.
```
Citation? Or at least elaboration.

seresistvanandras replied:
Added the references

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Therefore, the payment will fail.
+The question is: how exactly will it fail?
+
+There are two scenarios.
```
Simplified example, but this is missing a bit of nuance re the existence of the channel level flow control constraints that may be slightly out of sync with the advertised routing constraints. As an example, let's say a channel only allows 5 HTLCs for w/e reason in this case when a prober attempts to attach a 6th HTLC, they'll get a "temp chan failure" even though the channel may have had sufficient bandwidth to carry the channel. 

Another value to consider is the `max_htlc` setting, which may not allow a prober to deduce the full available bandwidth amount in the channel in a single go.

Roasbeef reviewed (changes requested):
Chunking my review a bit to avoid a massive dump of comments, will likely have 2/3 more review chunks to complete my final pass. 

My initial impression though is that this is an amazing document, and I don't think there exists another comparable approachable single piece of writing that attempts to approach security+privacy of LN from a similarly objective angle. Amazing contribution!

Trxlz commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+In summary, while some aspects of the Lightning Network's architecture suggest that it is a step forward in terms of privacy compared to Bitcoin, other properties of the protocol may make attacks on privacy easier.
+Thorough research is needed to evaluate what privacy guarantees the LN provides and improve the state of affairs.
+We are happy to report that multiple researchers teams are working on Lightning privacy.
```
```suggestion
We are happy to report that multiple research teams are working on Lightning privacy.
```

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+There are other potential countermeasures against channel probing.
+First, it is hard for an attacker to target unannounced channels.
+Second, nodes that implement JIT routing are less prone to the attack.
```
I think that should be relaxed to "may be" less prone to probing, as to my knowledge it hasn't really been deployed at a large scale to be able to draw solid empirical conclusions. It seems as though JIT routing would also exacerbate the issue of HTLC slot utilization, given that a single prob across `N` nodes would then cause each of those nodes to attempt to generate up to `K` routes themselves to attempt to rebalancing things, each of which may actually fail.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Similarly to Bitcoin, the Lightning Network charges fees for the use of its public resources, but in this case, the resources are public channels, and the fees come in the form of routing fees.
+The ability to route payments through nodes in exchange for fees provides the network with a large scalability benefit - nodes that are not directly connected can still transact - but it comes at the cost of exposing a public resource that must be protected against DoS attacks.
+
+When a Lightning node forwards a payment on your behalf, it uses bandwidth to update its commitment transaction, and the amount of the payment is reserved in their channel balance until it is settled or failed.
```
Unclear if "bandwidth" here refers to payment bandwidth, or actual data bandwidth, works in either interpretation though IMO.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+There are two known DoS attacks on public Lightning channels which render a target channel, or a set of target channels, unusable.
+Both attacks involve routing payments through a public channel, then holding them until their timeout, thus maximizing the attack's duration.
+The requirement to fail payments to not pay fees is fairly simple to meet because malicious nodes can simply reroute payments to themselves.
+In the absence of fees for failed payments, the only cost to the attacker is the on-chain cost of opening a channel to dispatch these payments through, which can be trivial in low fee environments.
```
The cost also includes worst case closing 2x the channels (if they're doing it via circular routes), and also the opportunity cost of the Bitcoin they need to put into channels to launch the attack.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Each party is limited to a total of https://github.com/lightningnetwork/lightning-rfc/blob/c053ce7afb4cbf88615877a0d5fc7b8dbe2b9ba0/02-peer-protocol.md#the-open_channel-message[483] on the commitment at a time.
+A channel jamming attack allows an attacker to render a channel unusable by routing 483 payments through the target channel and holding them until they timeout.
+
+It should be noted that this limit is arbitrary and was chosen in the specification to ensure that all the HTLCs can be swept in a https://github.com/lightningnetwork/lightning-rfc/blob/master/05-onchain.md#penalty-transaction-weight-calculation[single justice transaction].
```
It isn't arbitrary in that it was selected to respect the _policy_ limits on the max size of a transaction in the block.

seresistvanandras replied:
fixed

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+A channel jamming attack allows an attacker to render a channel unusable by routing 483 payments through the target channel and holding them until they timeout.
+
+It should be noted that this limit is arbitrary and was chosen in the specification to ensure that all the HTLCs can be swept in a https://github.com/lightningnetwork/lightning-rfc/blob/master/05-onchain.md#penalty-transaction-weight-calculation[single justice transaction].
+While this limit _may_ be increased, transactions are still limited by the block size, so the number of slots available is likely to remain limited.
```
One missing component here is that it's possible to essentially lift this limit (by using indirect commitments), instead letting nodes manage the free HTLC slots as if they were space in a queue. Modern implementations also feature something commonly referred to as an "HTLC interceptor" that allows them to make accept/deny decisions based on the current channel state. In a sense, the problem is similar to TCP congestion control/avoidance and buffer bloat itself.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+If the same IP or Tor addresses correspond to different LN nodes, these nodes are likely controlled by the same user.
+
+_Countermeasures_: For more privacy, aliases should be sufficiently different from one another.
+While the public announcement of IP addresses may be unavoidable for those nodes that wish to have incoming channels in the LN, linkability across nodes of the same user can be mitigated if the clients for each node are hosted with different service providers and thus IP addresses.
```
Tor also goes a long way here given that onion services addresses are effectively infinite, while IPv4 adresses are still scarce.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+On the other hand, the adversary could be more stealthy.
+Several topology-based attacks target a single node or a single payment channel.
+For example, an adversary might be interested in exhausting a certain payment channel's capacity on purpose.
+More generally, an adversary can deplete all the outgoing capacity of a node to knock it down from the routing market.
```
How so? Outgoing capacity is only "shifted around" and never really depleted in a sense. Also users are able to disable certain chanenls to protect the outgoing capacity of those channels.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+As a result, incentives drive graph development.
+Let's describe some of the relevant incentives:
+
+  * Rational incentives.
```
A new incentive comes in the form of Lightning Pool: aside from the _potential_ to earn routing fees, nodes are also able to earn passive yield simply by allocating capital to the network where it's most demanded.

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Many of the concerns listed in this chapter are likely to be addressed as it matures and grows.
+In the meantime, there are some measures that you can take to guard your node against malicious users; something as simple as updating the default parameters that your node runs with can go a long way in hardening your node.
+
+== Private channels
```
Another benefit is that one is able to more effectively guard/utilize its channel bandwidth: given that these private channels will only be used to receiving or sending directly to that node.

seresistvanandras replied:
Thanks! Added your insight!

Roasbeef commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+While mitigations are being developed on the protocol level, there are many steps that a node can take to protect against denial of service attacks on their public channels.
+
+* Minimum HTLC size: on channel open, your node can set the minimum HTLC size that it will accept.
+Setting a higher value ensures that each of your available channel slots cannot be occupied by a very small payment.
```
Another related setting here is `max_htlc`, which can be used to help mitigate probing attacks somewhat, and also let a node control the exposure it'll have to a given payment flow.

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+There are two scenarios.
+If Alice has more than 500 thousand satoshis on her side of the channel to Bob, she forwards the payment.
+Bob decrypts the payment onion and realizes that the payment is intended for him.
+He looks up his local store of payment secrets, searches for the preimage that corresponds to the payment hash but does not find one.
```
Use `pre-image` instead of `preimage` for consistency as the first version was used previously.

```suggestion
He looks up his local store of payment secrets, searches for the pre-image that corresponds to the payment hash but does not find one.
```

seresistvanandras replied:
Changed it to "preimage" everywhere.

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+This allows nodes to costlessly route failed payments through any channels.
+This is great for legitimate users, who wouldn't like to pay for failed attempts, but also allows attackers to costlessly consume nodes' resources - much like the low-fee transactions on Bitcoin that never end up paying miner fees.
+
+At the time of writing, a discussion is https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-June/002734.html[ongoing] on the lightning-dev mailing list as to how best address this issue.
```
Specifying timing twice is redundant. Also added a `to`.

```suggestion
A discussion is https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-June/002734.html[ongoing] on the lightning-dev mailing list as to how best to address this issue.
```

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+== Accepting channels
+At present, Lightning nodes struggle with bootstrapping inbound liquidity.
+This issue has lead to an attitude of accepting any channel that another peer attempts to open to your node.
```
tense

```suggestion
This issue has led to an attitude of accepting any channel that another peer attempts to open to your node.
```

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Integrity implies that the information does not get altered in transit.
+Availability ensures that users may expect the system to be functioning most of the time.
+
+The important properties of the Lightning Network are mostly centered around confidentiality and availability.
```
Why? Is is because integrity comes out of the box? If so, it's worth mentioning.

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Discovering someone's IP address or telephone number may also be considered deanonymization.
+A piece of information that allows linking a user's action to their previous actions is referred to as _identity_.
+Second, deanonymization is not binary.
+A user is not either fully anonymous or completely deanonymized.
```
```suggestion
A user is not fully anonymous nor completely deanonymized.
```

seresistvanandras replied:
Fixed

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+Bitcoin was initially perceived as an anonymous currency.
+Indeed, Bitcoin addresses are not linked to real identities.
+However, all transactions are broadcast in cleartext and can be analyzed.
+Multiple companies deanonymize users of Bitcoin and other cryptocurrencies.
```
```suggestion
Multiple companies were established to deanonymize users of Bitcoin and other cryptocurrencies.
```

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+While this improves the privacy baseline, other properties of the Lightning protocol may make anonymous payments more challenging.
+For instance, larger payments have fewer routing options.
+This may allow an adversary who controls well-capitalized nodes to route most large payments and discover payment amounts and probably other details.
+Another relevant difference between Lightning and Bitcoin is that Lightning nodes maintain a permanent identity, whereas Bitcoin nodes do not.
```
Although Bitcoin nodes do not assume a permanent identity, wallet clustering achieves similar outcomes for wallets without/under utilized/or ineffective privacy techniques.

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+A Lightning user, when sending a payment, has its neighbors in its anonymity set.
+Specifically, a routing node only knows the immediately preceding and following nodes.
+The routing node does not know whether its immediate neighbors in the payment route are the ultimate sender or receiver.
+Therefore, the anonymity set of a node in Lightning roughly equals its neighbors.
```
Not sure about this statement. Researchers are trying to fit the onion routing into their anonymity set (entropist) models with little success: https://www.freehaven.net/anonbib/cache/entropist.pdf

> Starting from an entropist conception of anonymity has led to a focus on system properties that are not the most important to system security, has led to system assumptions that are not reasonable in practice, has led to adversary models that are not reasonable in practice, and has led to both theory and system design for anonymous communication that fail in significant ways.

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+It then runs the attack sometime later at time _t'_ and uses the differences between the two snapshots to infer information about payments that took place by looking at paths that have changed.
+In the simplest case, if only a single payment occurred between time _t'_ and _t_, the adversary observes a single path where the balances have changed by the same amounts.
+Thus, the adversary learns everything about this payment: the sender, the recipient, and the amount.
+If multiple payment paths overlap, the adversary needs to apply heuristics to identify such overlap and separate the payments.
```
Since my LN knowledge is quite lacking, I can't claim to fully understand what's going on here, but I created some illustration to make it easier for the reader to grasp the basic concept at least. If you want to add it or create something similar incorporating the amount attack itself that could be helpful to improve my understanding without reading the rest of the book.
![image](https://user-images.githubusercontent.com/9156103/102006769-f6e0a800-3d23-11eb-8beb-0aef66543c5a.png)

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+
+Now, we turn our attention to an *on-path adversary*.
+Such an adversary might seem convoluted.
+However, the single most central node can already observe close to 50% of all LN payments, while the four most central nodes observe an average of 72% payments.
```
Specify when were these statistics are collected with citation as @Roasbeef suggested: https://github.com/lnbook/lnbook/pull/524/files#r537747008

Also if this information could be estimated from an already existing public website, then it could be useful to document how so in the future, the numbers could be easily updated.

seresistvanandras replied:
Added the appropriate references!

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+An on-path adversary can log for every routed payment along with the amount of time it takes for a node to respond to an HTLC request.
+Before starting the attack, the attacker learns every node's latency characteristics in the Lightning Network by sending them requests.
+Naturally, this can aid in establishing the adversary's precise position in the payment path.
+Even more, as it was recently shown, an attacker can successfully determine the sender and the receiver of a payment from a set of possible senders and receivers using time-based estimators.
```
Maybe make this statement timeless? You don't know when the reader will read this and unlike a research paper, this work evolves, so unless the reader has a git blame at hand he won't be able to tell what "recently" means.

```suggestion
Even more, an attacker can successfully determine the sender and the receiver of a payment from a set of possible senders and receivers using time-based estimators.
```

nopara73 commented on security_privacy_ln.asciidoc:
```diff
@@ -0,0 +1,614 @@
+She can continue probing, choosing the next probing amount such that it divides the current estimation interval in half.
+This well-known search technique is called _binary search_.
+With binary search, the number of probes is _logarithmic_ in the desired precision.
+For example, to obtain Alice's balance in a channel of 1 million satoshis up to a single satoshi, Mallory would only have to perform stem:[\log_2(1000000) \approx 20] probings.
```
This formula doesn't get previewed here: https://github.com/seresistvanandras/lnbook/blob/lnsecurityprivacy/security_privacy_ln.asciidoc#linking-senders-and-receivers

seresistvanandras replied:
Indeed, you're right. Locally, in my text editor, it renders correctly. Maybe, it's a fault of GitHub? Anyone could confirm this? cc @s-tikhomirov  @carlaKC

aantonop replied:
This is a Github rendering limitation, but our asciidoc rendered will handle it correctly

nopara73 reviewed (commented):
Great chapter, I've learnt a lot from it! General suggestions:

- It would greatly help my understanding if it'd have a lot more illustration.
- It reads more like a research paper, rather than a book. I'm not sure if it's consistent with the rest of the book.

seresistvanandras committed:
Fixing the misspelling of lead in past tense

Co-authored-by: nopara73 <adam.ficsor73@gmail.com>

seresistvanandras committed:
Typo fix

Co-authored-by: Trellz <70897306+Trxlz@users.noreply.github.com>

seresistvanandras committed:
Non-contentious fixes

seresistvanandras committed:
Non-contentious fixes 2

seresistvanandras committed:
Non-contentious fixes 3

seresistvanandras commented:
> Great chapter, I've learnt a lot from it! General suggestions:
> 
> * It would greatly help my understanding if it'd have a lot more illustration.

Absolutely agree. Would be nice if @Roasbeef @renepickhardt and @aantonop could tell us where it would be necessary to amend the text with more illustrations. Then, we could iterate on those figures right away before it gets too late.

ariard commented:
This is a great thread ! I wanna to do a review but I'm not sure about the breath/depth you're aiming for ?

The privacy section is pretty awesome. It's really thoughtful to talk about cross-layer linking (sadly you have far more heuristics than the ones already detailed...). Maybe a privacy framework should include LN nodes taxonomy (leafs, routing, merchants), it's likely you can observe a lot from their behavior patterns. Also risks of privacy leaks backfire on L1 coins, if utxos context-switches aren't well handled by your wallet.

On the security, it sounds limited to channel-jamming at the network level and doesn't mention other integrity concerns like channel/HTLC steals vectors, timevalue DoS, fees DoS, balance blackmailing, ... which affect link layer.


But ultimately, it's function of the public you're targeting ?
* node operators, who want to deploy their channels cautiously
* application devs, who want to inform their application design
* lightning devs, who have a hard time tracking the fast-evolving research
* base layer devs, who want a quick survey about L2 security & privacy
* curious users, who want to understand what LN is made of
* academic/security researchers, who want inspiration for their future work
* ...

And I guess that's only a small subset of the potential readers and their alleged motivations.

carlaKC commented:
> It reads more like a research paper, rather than a book. I'm not sure if it's consistent with the rest of the book.

Thanks for the feedback @nopara73! definitely something we want to address. Going to read through some of the other chapters to get an idea of tone/levelling.

> But ultimately, it's function of the public you're targeting ?

@ariad I'll defer to the main authors for authority, but from your list I would definitely rule out lightning+ base layer devs and researchers. Afaik, this book is not really intended for that level of specialization.

> doesn't mention other integrity concerns

In terms of other security concerns, it's difficult to outline them all in an approachable way, but I agree that this section could be fleshed out with flood&loot and a few other attacks. One concern I have is that some topics will be unfamiliar to a regular bitcoiner, for example transaction pinning, so we're going to have to go on a long tangent just to describe some of these attacks. 

This is already quite an content-dense chapter, so I think we're going to have to be satisfied with a disclaimer that the things we mention here are not exhaustive (privacy and security never can be, really).

aantonop commented:
> > But ultimately, it's function of the public you're targeting ?
> 
> @Ariad I'll defer to the main authors for authority, but from your list I would definitely rule out lightning+ base layer devs and researchers. Afaik, this book is not really intended for that level of specialization.
> 

The target audience for this book is a developer or computer science student. More specifically, technically-minded people with some background in Bitcoin who want to understand LN or develop applications that use LN. The first 4-5 chapters should also be accessible to non-developers but technical people with an interest in understanding the technology. Our target audience is not LN protocol developers or LN client developers.

aantonop commented:
What is the status of this PR? @s-tikhomirov ?

Have you addressed the comments from @Roasbeef ? Are we ready to merge this initial draft and subject it to further PRs for other editorial suggestions and changes?

seresistvanandras commented:
I only want to include a few cross-references across the chapter to increase readability. @Roasbeef requested a few clarifications on how certain inefficiencies and privacy leaks are mitigated by MPP payments. So, I think we are 98% ready for the merge. I'll add these latest this weekend hopefully.

Upon request, we would be happy to amend the accessibility of the chapter with graphs, figures, illustrations. However, we have not received yet any indications where such supplemental material would be useful, wanted. Let us know and then we can also add those figures and illustrations, if needed.

s-tikhomirov commented:
From my side, I have a deadline upcoming and can't allocate much time for this project for the next two weeks, unfortunately. I'll look at the state of this PR then and contribute if necessary.

nopara73 commented:
A new paper is out on LN privacy: https://mobile.twitter.com/GeorgeKappos/status/1352969635679920131

https://arxiv.org/abs/2003.12470

seresistvanandras committed:
LN and censorship resistance

Co-authored-by: nopara73 <adam.ficsor73@gmail.com>

seresistvanandras committed:
References to MPP

seresistvanandras committed:
Deanonymization is not trivial

Co-authored-by: nopara73 <adam.ficsor73@gmail.com>

seresistvanandras commented:
I think this PR is now ready to be merged. We addressed most of the comments and suggestions. Readability might be improved in the future by adding illustrations, figures and references to other chapters in the book.
